### PR TITLE
[dpc++] + ranges API for sort, stable_sort, merge algorithms (hetero versions), the tests

### PR DESCRIPTION
**Pull request by MikeDvorskiy**
1. Ranges API for sort, stable_sort, merge algorithms (hetero versions)
2. The minimal tests for ranges API sort, stable_sort, merge algorithms (hetero versions)

### CODE REVIEW

File: test/parallel_api/ranges/merge_ranges_sycl.pass.cpp [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#include "support/pstl_test_config.h"
+
```
Comment (evgeniusp) [inline test/parallel_api/ranges/merge_ranges_sycl.pass.cpp:17]: Do we really need this extra empty lines here and below?

File: test/parallel_api/ranges/merge_ranges_sycl.pass.cpp [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    using namespace TestUtils;
+    using namespace oneapi::dpl::experimental::ranges;
+    {
+        sycl::buffer<int> A(data1, sycl::range<1>(max_n));
+        sycl::buffer<int> B(data2, sycl::range<1>(max_n));
+        sycl::buffer<int> C(data3, sycl::range<1>(max_n));
+
+        auto exec = TestUtils::default_dpcpp_policy;
+        using Policy = decltype(TestUtils::default_dpcpp_policy);
```
Comment (evgeniusp) [inline test/parallel_api/ranges/merge_ranges_sycl.pass.cpp:45]: We don't use this alias in this file. Do we actually need this one?

File: test/parallel_api/ranges/merge_ranges_sycl.pass.cpp [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+#include "support/utils.h"
+
+#include <iostream>
+
+int32_t
+main()
+{
+#if _ONEDPL_USE_RANGES
+    const int max_n = 10;
+    int data1[max_n] = { 0,  1, 2, 3, 4, 5, 6, 6, 6, 6};
```
Comment (evgeniusp) [inline test/parallel_api/ranges/merge_ranges_sycl.pass.cpp:33]: Here, we have test-case with only non-empty arrays with fixed size. Do we need to add an additional test-cases with the different sizes (including empty) or non-range APIs cover it?

File: test/parallel_api/ranges/sort_ranges_sycl.pass.cpp [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#include "support/pstl_test_config.h"
+
```
Comment (evgeniusp) [inline test/parallel_api/ranges/sort_ranges_sycl.pass.cpp:17]: The same as in test for merge.

File: test/parallel_api/ranges/sort_ranges_sycl.pass.cpp [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+#include "support/utils.h"
+
+#include <iostream>
+
+int32_t
+main()
+{
+#if _ONEDPL_USE_RANGES
+    const int max_n = 10;
+    int data1[max_n] = {0, 1, 2, -1, 4, 5, 6, 7, 8, 9};
```
Comment (evgeniusp) [inline test/parallel_api/ranges/sort_ranges_sycl.pass.cpp:33]: The same as in test for merge.

File: test/parallel_api/ranges/stable_sort_ranges_sycl.pass.cpp [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    using namespace TestUtils;
+    using namespace oneapi::dpl::experimental::ranges;
+    {
+        sycl::buffer<int> A(data1, sycl::range<1>(max_n));
+        sycl::buffer<int> B(data2, sycl::range<1>(max_n));
+
+        auto exec = TestUtils::default_dpcpp_policy;
+        using Policy = decltype(TestUtils::default_dpcpp_policy);
+
+        stable_sort(exec, all_view<int, sycl::access::mode::read_write>(A));
```
Comment (evgeniusp) [inline test/parallel_api/ranges/stable_sort_ranges_sycl.pass.cpp:45]: The tests for sort and stable sort are the same with the exception of suitable algos calling (two strings). Maybe, we can combine it into one test. What do you think?

File: include/oneapi/dpl/pstl/glue_algorithm_ranges_defs.h [hunk 7]
```diff
@@ -125,6 +145,18 @@ template <typename _ExecutionPolicy, typename _Range>
 oneapi::dpl::__internal::__enable_if_execution_policy<_ExecutionPolicy, bool>
 is_sorted(_ExecutionPolicy&& __exec, _Range&& __rng);
 
+// [alg.merge]
+
+template <typename _ExecutionPolicy, typename _Range1, typename _Range2, typename _Range3, typename _Compare>
+oneapi::dpl::__internal::__enable_if_execution_policy<_ExecutionPolicy,
+                                                      oneapi::dpl::__internal::__difference_t<_Range3>>
```
[the remark this answers is not in the document]
Comment (MikeDvorskiy) [reply include/oneapi/dpl/pstl/glue_algorithm_ranges_defs.h:152]: Range based API algo return  a difference type instead of OutputIterator, to avoid introduce type of "OutputIterator", which is compatible with a corresponding range.
Actually, there are no  parallel versions of range based API algorithm in the standard yet. Currently we have added a kind of extensions.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_fpga.h [hunk 8]
```diff
@@ -216,15 +216,16 @@ __parallel_merge(_ExecutionPolicy&& __exec, _Range1&& __rng1, _Range2&& __rng2,
...
+__parallel_stable_sort(_ExecutionPolicy&& __exec, _Range&& __rng, _Compare __comp)
     -> oneapi::dpl::__internal::__enable_if_fpga_execution_policy<
-        _ExecutionPolicy, decltype(oneapi::dpl::__par_backend_hetero::__parallel_stable_sort(__device_policy(__exec),
-                                                                                             __first, __last, __comp))>
+        _ExecutionPolicy, decltype(oneapi::dpl::__par_backend_hetero::__parallel_stable_sort(
+                              __device_policy(__exec), ::std::forward<_Range>(__rng), __comp))>
 {
     // workaround until we implement more performant version for patterns
-    return oneapi::dpl::__par_backend_hetero::__parallel_stable_sort(__device_policy(__exec), __first, __last, __comp);
+    oneapi::dpl::__par_backend_hetero::__parallel_stable_sort(__device_policy(__exec), ::std::forward<_Range>(__rng),
```
Comment (dsobolev-dev) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_fpga.h:227]: Please, add `return` statement since we must provide a `__future` object to the caller.

File: test/parallel_api/ranges/merge_ranges_sycl.pass.cpp [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    int data2[max_n] = {-2, -1, 2, 3, 3, 5, 6, 7, 8, 9};
+    int data3[max_n] = { 0,  0, 0, 0, 0, 0, 0, 0, 0, 0};
```
Comment (dsobolev-dev) [inline test/parallel_api/ranges/merge_ranges_sycl.pass.cpp:35]:
```suggestion
    int data3[2 * max_n] = {};
```
This will initialize `data3` with zeroes (https://en.cppreference.com/w/cpp/language/zero_initialization: _If T is array type, each element is zero-initialized_). We need `2 * max_n` for merging.

File: test/parallel_api/ranges/merge_ranges_sycl.pass.cpp [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+        sycl::buffer<int> C(data3, sycl::range<1>(max_n));
+
+        auto exec = TestUtils::default_dpcpp_policy;
+        using Policy = decltype(TestUtils::default_dpcpp_policy);
+
+        merge(exec, all_view(A), all_view(B), all_view<int, sycl::access::mode::write>(C));
+    }
+
+    //check result
+    bool res = ::std::is_sorted(data3, data3 + max_n);
```
[the remark this answers is not in the document]
Comment (dsobolev-dev) [reply test/parallel_api/ranges/merge_ranges_sycl.pass.cpp:51]: You can also use `std::includes` as it is done in `merge.pass`

File: test/parallel_api/ranges/merge_ranges_sycl.pass.cpp (hunk 3 above)
Comment (MikeDvorskiy) [reply test/parallel_api/ranges/merge_ranges_sycl.pass.cpp:33]: The non-range APIs (to be more precise  - hetero patterns) cover it.

File: test/parallel_api/ranges/sort_ranges_sycl.pass.cpp (hunk 5 above)
Comment (MikeDvorskiy) [reply test/parallel_api/ranges/sort_ranges_sycl.pass.cpp:33]: see my comment above.

File: test/parallel_api/ranges/merge_ranges_sycl.pass.cpp (hunk 1 above)
Comment (MikeDvorskiy) [reply test/parallel_api/ranges/merge_ranges_sycl.pass.cpp:17]: it is a kind of visual separator..

File: test/parallel_api/ranges/merge_ranges_sycl.pass.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (MikeDvorskiy) [reply test/parallel_api/ranges/merge_ranges_sycl.pass.cpp:51]: As far as we test just Range API here,
such check "bool res = ::std::is_sorted(data3, data3 + 2 * max_n);" is enough, I guess.

File: test/parallel_api/ranges/sort_ranges_sycl.pass.cpp [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+    {
+        sycl::buffer<int> A(data1, sycl::range<1>(max_n));
+        sycl::buffer<int> B(data2, sycl::range<1>(max_n));
+
+        auto exec = TestUtils::default_dpcpp_policy;
+        using Policy = decltype(TestUtils::default_dpcpp_policy);
+
+        sort(exec, all_view<int, sycl::access::mode::read_write>(A));
+        sort(make_new_policy<new_kernel_name<Policy, 0>>(exec), all_view<int, sycl::access::mode::read_write>(B),
+            oneapi::dpl::__internal::__pstl_greater());
```
[the remark this answers is not in the document]
Comment (MikeDvorskiy) [reply test/parallel_api/ranges/sort_ranges_sycl.pass.cpp:47]: We test just  range API here. There is the good coverage  in iterator-based tests which are based on the same sort pattern.

File: test/parallel_api/ranges/merge_ranges_sycl.pass.cpp [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+    using namespace TestUtils;
+    using namespace oneapi::dpl::experimental::ranges;
+    {
+        sycl::buffer<int> A(data1, sycl::range<1>(max_n));
+        sycl::buffer<int> B(data2, sycl::range<1>(max_n));
+        sycl::buffer<int> C(data3, sycl::range<1>(max_n));
+
+        auto exec = TestUtils::default_dpcpp_policy;
+
+        merge(exec, all_view(A), all_view(B), all_view<int, sycl::access::mode::write>(C), oneapi::dpl::__internal::__pstl_less());
```
Comment (dsobolev-dev) [inline test/parallel_api/ranges/merge_ranges_sycl.pass.cpp:46]: For better code coverage, we should test both newly introduced APIs: with a predicate and without it.

File: include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h [hunk 13]
```diff
@@ -286,6 +294,63 @@ __pattern_adjacent_find(_ExecutionPolicy&& __exec, _Range&& __rng, _BinaryPredic
...
+
+template <typename _ExecutionPolicy, typename _Range1, typename _Range2, typename _Range3, typename _Compare>
+oneapi::dpl::__internal::__enable_if_hetero_execution_policy<_ExecutionPolicy,
+                                                             oneapi::dpl::__internal::__difference_t<_Range3>>
+__pattern_merge(_ExecutionPolicy&& __exec, _Range1&& __rng1, _Range2&& __rng2, _Range3&& __rng3, _Compare __comp)
+{
+    auto __n1 = __rng1.size();
+    auto __n2 = __rng2.size();
+    auto __n = __n1 + __n2;
+    if (__n == 0)
```
Comment (dsobolev-dev) [inline include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h:309]: Please note that the following test case:
```c++
#include "support/pstl_test_config.h"

#include _PSTL_TEST_HEADER(execution)
#if _ONEDPL_USE_RANGES
#include _PSTL_TEST_HEADER(ranges)
#endif

#include "support/utils.h"
#include <iostream>

int32_t
main()
{
#if _ONEDPL_USE_RANGES
    using T = int;

    std::vector<T> in;

    using namespace TestUtils;
    using namespace oneapi::dpl::experimental::ranges;
    {
        sycl::buffer<T> C(in.data(), sycl::range<1>(in.size()));
        std::cout << "all_view(C).size(): " << all_view(C).size() << std::endl;
    }
#endif //_ONEDPL_USE_RANGES
    ::std::cout << TestUtils::done() << ::std::endl;
    return 0;
}
```
fails with:
```c++
terminate called after throwing an instance of 'cl::sycl::invalid_object_error'
  what():  SYCL buffer size is zero. To create a device accessor, SYCL buffer size must be greater than zero. -30 (CL_INVALID_VALUE)
```

That means that the following checks will never trigger because we are not able to create a zero-sized range from a zero-sized buffer.

Comment (dsobolev-dev) [reply include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h:309]: This may be a more general issue, thus not a subject of this patch.

File: test/parallel_api/ranges/merge_ranges_sycl.pass.cpp [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+    using namespace oneapi::dpl::experimental::ranges;
+    {
+        sycl::buffer<int> A(data1, sycl::range<1>(max_n));
+        sycl::buffer<int> B(data2, sycl::range<1>(max_n));
+        sycl::buffer<int> C(data3, sycl::range<1>(max_n));
+
+        auto exec = TestUtils::default_dpcpp_policy;
+        using Policy = decltype(TestUtils::default_dpcpp_policy);
+
+        merge(exec, all_view(A), all_view(B), all_view<int, sycl::access::mode::write>(C));
```
[the remark this answers is not in the document]
Comment (dsobolev-dev) [reply test/parallel_api/ranges/merge_ranges_sycl.pass.cpp]: For better code coverage, we should test both newly introduced APIs: with a predicate and without it.
You may consider the following test (only main function): 
```c++
int32_t
main()
{
#if _ONEDPL_USE_RANGES
    using T = int;

    const int in_n = 10;
    const int out_n = 2 * in_n;

    T in1[in_n] = { 0,  1, 2, 3, 4, 5, 6, 6, 6, 6};
    T in2[in_n] = {-2, -1, 2, 3, 3, 5, 6, 7, 8, 9};
    T out1[out_n] = {};
    T out2[out_n] = {};

    using namespace TestUtils;
    using namespace oneapi::dpl::experimental::ranges;
    {
        sycl::buffer<T> A(in1, sycl::range<1>(in_n));
        sycl::buffer<T> B(in2, sycl::range<1>(in_n));
        sycl::buffer<T> D(out1, sycl::range<1>(out_n));
        sycl::buffer<T> E(out2, sycl::range<1>(out_n));

        auto exec = TestUtils::default_dpcpp_policy;

        merge(exec, all_view(A), all_view(B), all_view<T, sycl::access::mode::write>(D));
        merge(exec, all_view(A), all_view(B), all_view<T, sycl::access::mode::write>(E), ::std::less<T>());
    }

    //check result
    bool res1 = ::std::is_sorted(out1, out1 + out_n, ::std::less<T>());
    res1 &= ::std::includes(out1, out1 + out_n, in1, in1 + in_n, ::std::less<T>());
    res1 &= ::std::includes(out1, out1 + out_n, in2, in2 + in_n, ::std::less<T>());
    EXPECT_TRUE(res1, "wrong effect from 'merge' with sycl ranges");

    bool res2 = ::std::is_sorted(out2, out2 + out_n, ::std::less<T>());
    res2 &= ::std::includes(out2, out2 + out_n, in1, in1 + in_n, ::std::less<T>());
    res2 &= ::std::includes(out2, out2 + out_n, in2, in2 + in_n, ::std::less<T>());
    EXPECT_TRUE(res2, "wrong effect from 'merge' with sycl ranges with predicate");
#endif //_ONEDPL_USE_RANGES

    ::std::cout << TestUtils::done() << ::std::endl;
    return 0;
}

```

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_fpga.h (hunk 8 above)
Comment (dsobolev-dev) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_fpga.h:227]: Without that statement we will get the following issue on FPGA emulator:
```c++
parallel_backend_sycl_fpga.h:229:1: warning: non-void function does not return a value [-Wreturn-type]
...
2 warnings generated.
./sort_ranges_sycl.pass.exe
terminate called after throwing an instance of 'std::system_error'
  what():  Invalid argument
make: *** [Makefile:91: sort_ranges_sycl.pass] Aborted (core dumped)
```

File: test/parallel_api/ranges/sort_ranges_sycl.pass.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (dsobolev-dev) [reply test/parallel_api/ranges/sort_ranges_sycl.pass.cpp:47]: I guess we can easily change it to `::std::greater<int>()`. It looks simpler too. But if you think it is not necessary, then leave it as is.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_fpga.h (hunk 8 above)
Comment (dsobolev-dev) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_fpga.h:227]:
```suggestion
    return oneapi::dpl::__par_backend_hetero::__parallel_stable_sort(__device_policy(__exec), ::std::forward<_Range>(__rng),
```

File: test/parallel_api/ranges/merge_ranges_sycl.pass.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (dsobolev-dev) [reply test/parallel_api/ranges/merge_ranges_sycl.pass.cpp]: Discussed offline with @MikeDvorskiy who agreed to use `std::includes` to make test more reliable as well as not to use internal predicates.

File: include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h (hunk 13 above)
Comment (MikeDvorskiy) [reply include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h:309]: About "if (__n == 0)" and "never trigged"  - essentially, we may have a range which is based on a range factory(iota, etc), non-sycl_view over sycl buffer.

**Review by dsobolev-dev — APPROVED** [review]
Testing system (which is done via cmake) does not support separate execution of different test cases from one file, that is why I think we can leave different files for `sort` and `stable_sort` for now. 
The patch looks good to me.

File: test/parallel_api/ranges/stable_sort_ranges_sycl.pass.cpp (hunk 6 above)
Comment (MikeDvorskiy) [reply test/parallel_api/ranges/stable_sort_ranges_sycl.pass.cpp:45]: For the folder "ranges" we have one test case - one test file. It's pretty simple for statistic. 
Yes, we have some code duplication.. but the tests are so trivial.. I'm not sure that we have to combine several tests into one in these test units.
